fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/PipelineDriverAllocator.java

```diff
@@ -0,0 +1,184 @@
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Calculate proper pipeline DOP for adaptive DOP query according to the fragments and DOP of running queries.
+ *
+ * <p> It is effective only when {@link GlobalVariable#isQueryQueueDriverHighWaterEffective()} returns true
+ * and pipeline_dop is 0 for query.
+ */
+public class PipelineDriverAllocator {
+    private static final Logger LOG = LogManager.getLogger(PipelineDriverAllocator.class);
+
+    private final Set<TUniqueId> allocatedSlotIds = Sets.newConcurrentHashSet();
+    private final AtomicInteger numAllocatedDrivers = new AtomicInteger();
+
+    private final Queue<AllocationRequest> allocationRequests = Queues.newConcurrentLinkedQueue();
+
+    @VisibleForTesting
+    static final int NUM_BATCH_SLOTS = 1024;
+
+    private static class AllocationRequest {
+        private final LogicalSlot slot;
+        private final CompletableFuture<Void> doneFuture = new CompletableFuture<>();
+
+        public AllocationRequest(LogicalSlot slot) {
+            this.slot = slot;
+        }
+
+        public void waitFinish() {
+            try {
+                doneFuture.get();
+            } catch (InterruptedException | ExecutionException e) {
+                LOG.warn("[Slot] wait allocation finish failed [slot={}]", slot, e);
+                Thread.currentThread().interrupt();
+            }
+        }
+
+        public void finish() {
+            doneFuture.complete(null);
+        }
+    }
+
+    public void allocate(LogicalSlot slot) {
+        if (!GlobalVariable.isQueryQueueDriverHighWaterEffective()) {
+            return;
+        }
+
+        boolean hasAllocated = !allocatedSlotIds.add(slot.getSlotId());
+        if (hasAllocated) {
+            return;
+        }
+
+        if (!slot.isAdaptiveDop()) {
+            numAllocatedDrivers.getAndAdd(slot.getNumDrivers());
+            return;
+        }
+
+        // If there are too many running drivers, set DOP to 1.
+        if (numAllocatedDrivers.get() + slot.getNumFragments() >= GlobalVariable.getQueryQueueDriverHighWater()) {
+            slot.setPipelineDop(1);
+            numAllocatedDrivers.getAndAdd(slot.getNumDrivers());
+            return;
+        }
+
+        AllocationRequest allocationRequest = new AllocationRequest(slot);
+        allocationRequests.add(allocationRequest);
+        allocateInBatch();
+        allocationRequest.waitFinish();
+    }
+
+    public void release(LogicalSlot slot) {
+        boolean hasAllocated = slot != null && allocatedSlotIds.remove(slot.getSlotId());
+        if (hasAllocated) {
+            numAllocatedDrivers.getAndAdd(-slot.getNumDrivers());
+        }
+    }
+
+    @VisibleForTesting
+    int getNumAllocatedDrivers() {
+        return numAllocatedDrivers.get();
+    }
+
+    private void allocateInBatch() {
```

Comment:
This implementation looks like WriteBatch merging in leveldb.

Reply:
Yes, inspiration from leveldb.